quaddtype/numpy_quaddtype/src/casts.cpp

```diff
@@ -483,38 +507,107 @@ quad_to_unicode_loop_aligned(PyArrayMethod_Context *context, char *const data[],
     npy_intp out_stride = strides[1];
 
     PyArray_Descr *const *descrs = context->descriptors;
-    QuadPrecDTypeObject *descr_in = (QuadPrecDTypeObject *)descrs[0];
-    QuadBackendType backend = descr_in->backend;
+    QuadPrecDTypeObject *descr_out = (QuadPrecDTypeObject *)descrs[1];
+    QuadBackendType backend = descr_out->backend;
 
-    npy_intp unicode_size_chars = descrs[1]->elsize / 4;
+    npy_intp bytes_size = descrs[0]->elsize;
 
     while (N--) {
-        quad_value in_val;
-        if (backend == BACKEND_SLEEF) {
-            in_val.sleef_value = *(Sleef_quad *)in_ptr;
+        quad_value out_val;
+        
+        if (bytes_to_quad_convert(in_ptr, bytes_size, backend, &out_val) < 0) {
+            return -1;
         }
-        else {
-            in_val.longdouble_value = *(long double *)in_ptr;
+
+        store_quad<Aligned>(out_ptr, out_val, backend);
+
+        in_ptr += in_stride;
+        out_ptr += out_stride;
+    }
+
+    return 0;
+}
+
+// QuadDType to bytes
+static NPY_CASTING
+quad_to_bytes_resolve_descriptors(PyObject *NPY_UNUSED(self), PyArray_DTypeMeta *dtypes[2],
+                                   PyArray_Descr *given_descrs[2], PyArray_Descr *loop_descrs[2],
+                                   npy_intp *view_offset)
+{
+    npy_intp required_size_bytes = QUAD_STR_WIDTH;
+
+    if (given_descrs[1] == NULL) {
+        PyArray_Descr *new_descr = PyArray_DescrNewFromType(NPY_STRING);
+        if (new_descr == NULL) {
+            return (NPY_CASTING)-1;
         }
+        new_descr->elsize = required_size_bytes;
+        loop_descrs[1] = new_descr;
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
 
-        // Convert to Sleef_quad for Dragon4
-        Sleef_quad sleef_val = quad_to_sleef_quad(&in_val, backend);
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
 
-        // Get string representation with adaptive notation
-        PyObject *py_str = quad_to_string_adaptive(&sleef_val, unicode_size_chars);
+    *view_offset = 0;
+
+    // If target descriptor is wide enough, it's a safe cast
+    if (loop_descrs[1]->elsize >= required_size_bytes) {
+        return NPY_SAFE_CASTING;
+    }
+    return NPY_SAME_KIND_CASTING;
+}
+
+// Helper function: Copy string to bytes output buffer
+static inline void
+copy_string_to_bytes(const char *str, char *out_bytes, npy_intp bytes_size)
```

Comment:
I'd call this `copy_cstring_to_bytes` to make it clearer that `str` has to be null-terminated.

Reply:
or just inline it because the function only has one caller